### PR TITLE
Nc/runnable lambda recurse

### PR DESCRIPTION
**Pull request by nfcampos**

### CODE REVIEW

**Comment by hinthornw** [comment]
Guess this will be more visible after rebasing

File: libs/langchain/langchain/schema/runnable/base.py [hunk 1]
```diff
@@ -1502,17 +1514,87 @@ def __init__(self, func: Callable[[Input], Output]) -> None:
...
+    async def ainvoke(
+        self,
+        input: Input,
+        config: Optional[RunnableConfig] = None,
+        **kwargs: Optional[Any],
+    ) -> Output:
+        if hasattr(self, "afunc"):
+            return await self._acall_with_config(self._ainvoke, input, config)
+        else:
+            return await super().ainvoke(input, config)
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/base.py:1597]:
```suggestion
            # Delegating to super implementation of ainvoke.
            # Uses asyncio executor to run the sync version (invoke) 
            return await super().ainvoke(input, config)
```

File: libs/langchain/langchain/schema/runnable/base.py [hunk 2]
```diff
@@ -1618,10 +1700,12 @@ async def atransform(
 
 
 def patch_config(
-    config: RunnableConfig, callback_manager: BaseCallbackManager
+    config: RunnableConfig, callbacks: BaseCallbackManager, **kwargs: Any
 ) -> RunnableConfig:
     config = config.copy()
-    config["callbacks"] = callback_manager
+    config["callbacks"] = callbacks
+    # Correct typing requires Unpack, which isn't supported in all Python versions
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/base.py:1707]: I think typing_extensions can be installed to use Unpack

File: libs/langchain/tests/unit_tests/schema/test_runnable.py [hunk 3]
```diff
@@ -409,6 +410,50 @@ async def test_prompt_with_llm(
     )
 
 
+@pytest.mark.asyncio
```
Comment (eyurtsev) [inline libs/langchain/tests/unit_tests/schema/test_runnable.py:413]: Could we add some simple tests with runnable lambdas? 

```
def _simple_recursion(x: int) -> Union[int, Runnable]:
    if x < 10:
        return RunnableLambda(lambda *args: _simple_recursion(x + 1))
    else:
        return x


def test_recursive_lambda() -> None:
    runnable = RunnableLambda(_simple_recursion)
    assert runnable.invoke(5) == 10
```

and something to test recursion limit?

File: libs/langchain/langchain/schema/runnable/base.py [hunk 4]
```diff
@@ -1491,9 +1492,20 @@ class RunnableLambda(Runnable[Input, Output]):
     A runnable that runs a callable.
     """
 
-    def __init__(self, func: Callable[[Input], Output]) -> None:
-        if callable(func):
-            self.func = func
+    recursion_limit: int = 10
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/base.py:1495]: Can we pass this using init?

File: libs/langchain/langchain/schema/runnable/base.py [hunk 5]
```diff
@@ -1502,17 +1514,87 @@ def __init__(self, func: Callable[[Input], Output]) -> None:
...
+                    config, run_manager.get_child(), recursion_limit=recursion_limit - 1
+                ),
+            )
+        return output
+
+    async def _ainvoke(
+        self,
+        input: Input,
+        run_manager: AsyncCallbackManagerForChainRun,
+        config: RunnableConfig,
```
Comment (eyurtsev) [inline libs/langchain/langchain/schema/runnable/base.py:1554]: does runnable config have recursion_limit?

File: libs/langchain/langchain/schema/runnable/base.py (hunk 2 above)
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/base.py:1709]: No, it requires a Mypy flag

File: libs/langchain/tests/unit_tests/schema/test_runnable.py (hunk 3 above)
Comment (nfcampos) [reply libs/langchain/tests/unit_tests/schema/test_runnable.py]: Done

File: libs/langchain/langchain/schema/runnable/base.py (hunk 5 above)
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/base.py:1404]: Moved to config

File: libs/langchain/langchain/schema/runnable/base.py (hunk 4 above)
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/base.py]: Moved to config

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**